---
title: "Infer the secreted protein activity from gene expression in bulk level"
output: rmarkdown::html_vignette
description: >
  Infer the secreted protein activity from gene expression in bulk level.
vignette: >
  %\VignetteIndexEntry{Infer the secreted protein activity from gene expression in bulk level}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
---
output: github_document
---


```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This tutorial demonstrates how to run SecAct to infer secreted protein activity in bulk level. The input expression values, from either RNASeq or MicroArray, should be transformed by log2(x+1). x could be FPKM, RPKM, or TPM for RNASeq. Moreover, the input expression matrix should be normalized with the control samples.

## Case 1 

The current case includes the differential expression profiles between anti-IFNG treatment and control samples.

``` r
library(SecAct)

# prepare expression matrix
exprPath <- file.path(system.file(package="SecAct"), "extdata/IFNG_GSE100093.diff.gz")
expr <- read.csv(exprPath, row.names=1, check.names=F)

# run SecAct to infer activity
res <- SecAct.inference(expr, lambda=10000, nrand=1000)

# show activity
head(res$zscore)
``` 

## Case 2

This case aims to infer secreted protein activity for pancreatic cancer RNA-Seq cohorts from CPTAC.

``` r
library(SecAct)

# read expression matrix
exprPath <- file.path(system.file(package="SecAct"), "extdata/CPTAC_PDAC_Tumor.csv.gz")
expr <- read.csv(exprPath, row.names=1, check.names=F)

exprControlPath <- file.path(system.file(package="SecAct"), "extdata/CPTAC_PDAC_Normal.csv.gz")
exprControl <- read.csv(exprControlPath, row.names=1, check.names=F)

# normalize expression matrix
expr <- expr - rowMeans(exprControl)

# run SecAct to infer activity
res <- SecAct.inference(expr, lambda=10000, nrand=1000)

# show activity
head(res$zscore)
``` 

Note: if you do not have control samples, we recommend to normalize the expression values of each gene to zero mean across samples. In other words, just use the same expression matrix as control.

``` r
# normalize the expression values of each gene to zero mean across samples
expr <- expr - rowMeans(expr)
``` 
