<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Clinical relevance of secreted proteins in a large patient cohort • SecAct</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Clinical relevance of secreted proteins in a large patient cohort">
<meta name="description" content="Clinical relevance of secreted proteins in a large patient cohort.
">
<meta property="og:description" content="Clinical relevance of secreted proteins in a large patient cohort.
">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SecAct</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fas fa-home fa-lg"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials">
<li><h6 class="dropdown-header" data-toc-skip>Spatial transcriptomcis (ST) data</h6></li>
    <li><a class="dropdown-item" href="../articles/stPattern.html">Signaling patterns and velocities for multi-cellular ST data</a></li>
    <li><a class="dropdown-item" href="../articles/stCCC.html">Intercellular communication for single-cell resolution ST data</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Single-cell RNA sequencing data</h6></li>
    <li><a class="dropdown-item" href="../articles/scState.html">Secreted protein signaling activity for various cell states</a></li>
    <li><a class="dropdown-item" href="../articles/scCCC.html">Cell-cell communication mediated by secreted proteins</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bulk RNA sequencing data</h6></li>
    <li><a class="dropdown-item" href="../articles/bulkCohort.html">Clinical relevance of secreted proteins in a large patient cohort</a></li>
    <li><a class="dropdown-item" href="../articles/bulkChange.html">Secreted protein activity change between two phenotypes</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/data2intelligence/SecAct/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Clinical relevance of secreted proteins in a large patient cohort</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/data2intelligence/SecAct/blob/HEAD/vignettes/bulkCohort.Rmd" class="external-link"><code>vignettes/bulkCohort.Rmd</code></a></small>
      <div class="d-none name"><code>bulkCohort.Rmd</code></div>
    </div>

    
    
<p>This tutorial demonstrates how to infer the signaling activity of
secreted proteins from a large data cohort, and then calculate the
relevance of their activity to clinical variables. The input expression
values, from either RNA sequencing or MicroArray, should be transformed
by log2(x+1). Of note, x could be FPKM, RPKM, or TPM for RNA-Seq
data.</p>
<p>For this demonstration, we use an immunotherapy
<a href="https://github.com/ParkerICI/prince-trial-data" target="_blank" class="external-link">cohort</a>
of patients with pancreatic cancer. Using the SecAct framework, we
identify secreted proteins whose activity levels are associated with
clinical outcomes.</p>
<div class="section level2">
<h2 id="prepare-expression-data">Prepare expression data<a class="anchor" aria-label="anchor" href="#prepare-expression-data"></a>
</h2>
<p>This section shows how to read an expression matrix to the R
environment.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/data2intelligence/SecAct" class="external-link">SecAct</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># prepare expression matrix</span></span>
<span><span class="va">dataPath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"SecAct"</span><span class="op">)</span>, <span class="st">"extdata/"</span><span class="op">)</span></span>
<span><span class="va">expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dataPath</span>,<span class="st">"Pancreatic_Nivolumab_Padron2022.logTPM.gz"</span><span class="op">)</span>, check.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="infer-secreted-protein-activity">Infer secreted protein activity<a class="anchor" aria-label="anchor" href="#infer-secreted-protein-activity"></a>
</h2>
<p>We use <code>SecAct.activity.inference</code> to infer the signaling
activity of secreted proteins across patients.</p>
<p>If your data have control samples (e.g., normal patients), you can
assign them to <code>inputProfile_control</code>. <code>SecAct</code>
would normalize each gene’s expression values with the control samples,
i.e., subtracting the mean gene expression of control samples.</p>
<p>If not available, just set <code>inputProfile_control=NULL</code>
(default). <code>SecAct</code> would normalize the expression values of
each gene to zero mean across all input profiles. In other words, the
mean gene expression of all input profiles are regarded as control.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># infer activity; ~2 mins</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SecAct.activity.inference.html">SecAct.activity.inference</a></span><span class="op">(</span>inputProfile<span class="op">=</span><span class="va">expr</span>, inputProfile_control<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># res$zscore stores activity</span></span>
<span><span class="va">act</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">zscore</span></span>
<span></span>
<span><span class="co"># show activity</span></span>
<span><span class="va">act</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<p>The inferred protein activity is a relative measure. A positive value
indicates high activity in the corresponding patients, while a negative
value signifies low activity.</p>
</div>
<div class="section level2">
<h2 id="calculate-clinical-relevance">Calculate clinical relevance<a class="anchor" aria-label="anchor" href="#calculate-clinical-relevance"></a>
</h2>
<p>We next calculate the risk score for each secreted protein by linking
their activity levels with clinical information. For the clinical file,
please make sure the name of the 1st and 2nd columns are “Time” and
“Event”, respectively. <code>SecAct.coxph.regression</code> will conduct
Cox proportional hazard (PH) regression, with risk scores represented as
z-scores (Coef / StdErr) from the two-sided Wald test. Of note, clinical
covariates (e.g., age, gender, stage) would also be considered in
regression if available.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read clinical infomation</span></span>
<span><span class="va">clinical</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dataPath</span>,<span class="st">"Pancreatic_Nivolumab_Padron2022.OS_Nivo+Sotiga+Chemo"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##    Time Event Age Gender ECOG</span></span>
<span><span class="co">## 8   834     0  55      1    0</span></span>
<span><span class="co">## 13   28     1  60      0    1</span></span>
<span><span class="co">## 15  273     1  69      0    0</span></span>
<span><span class="co">## 22  159     1  69      0    1</span></span>
<span><span class="co">## 24  569     0  62      0    0</span></span>
<span><span class="co">## 25  264     1  66      1    1</span></span>
<span></span>
<span><span class="co"># compute risk score</span></span>
<span><span class="va">riskMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SecAct.coxph.regression.html">SecAct.coxph.regression</a></span><span class="op">(</span><span class="va">act</span>, <span class="va">clinical</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">riskMat</span><span class="op">)</span></span></code></pre></div>
<p>For the risk score z, a positive value means the secreted protein is
associated with poorer survival while a negative value means the
secreted protein is associated with better survival.</p>
</div>
<div class="section level2">
<h2 id="visualize-risk-score">Visualize risk score<a class="anchor" aria-label="anchor" href="#visualize-risk-score"></a>
</h2>
<p>User can visualize any secreted proteins of interest using
<code>SecAct.lollipop.plot</code>. Here, we select the top high and low
risk secreted proteins (SPs), which are associated with poorer and
better survival, respectively.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select the top secreted proteins</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span><span class="va">high.risk.SPs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">riskMat</span><span class="op">[</span>,<span class="st">"risk score z"</span><span class="op">]</span>,decreasing<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n</span><span class="op">]</span></span>
<span><span class="va">low.risk.SPs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">riskMat</span><span class="op">[</span>,<span class="st">"risk score z"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n</span><span class="op">]</span></span>
<span></span>
<span><span class="va">SPs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">high.risk.SPs</span>, <span class="va">low.risk.SPs</span><span class="op">)</span></span>
<span><span class="co"># user can assign any secreted proteins to SPs</span></span>
<span></span>
<span><span class="co"># generate a vector</span></span>
<span><span class="va">fg.vec</span> <span class="op">&lt;-</span> <span class="va">riskMat</span><span class="op">[</span><span class="va">SPs</span>,<span class="st">"risk score z"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># visualize in lollipop plot</span></span>
<span><span class="fu"><a href="../reference/SecAct.lollipop.plot.html">SecAct.lollipop.plot</a></span><span class="op">(</span><span class="va">fg.vec</span>, title<span class="op">=</span><span class="st">"Risk score z"</span><span class="op">)</span></span></code></pre></div>
<p><img src="img/bulk_RiskScore.png" width="50%" style="border:none"></p>
</div>
<div class="section level2">
<h2 id="draw-survival-plot">Draw survival plot<a class="anchor" aria-label="anchor" href="#draw-survival-plot"></a>
</h2>
<p>User can select a secreted protein of interest to draw its survival
curve using <code>SecAct.survival.plot</code>. The activity cutoff was
selected by maximizing the difference between high and low activity
patient groups.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Take secreted protein WNT7B as an example</span></span>
<span><span class="va">riskMat</span><span class="op">[</span><span class="st">"WNT7B"</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># visualize in survival plot</span></span>
<span><span class="fu"><a href="../reference/SecAct.survival.plot.html">SecAct.survival.plot</a></span><span class="op">(</span><span class="va">act</span>, <span class="va">clinical</span>, <span class="st">"WNT7B"</span>, x.title<span class="op">=</span><span class="st">"Overall (Days)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="img/bulk_Survival.png" width="50%" style="border:none"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Beibei Ru.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
