<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Clinical relavance of secreted proteins in a large patient cohort.
">
<title>Clinical relavance of secreted proteins in a large patient cohort • SecAct</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Clinical relavance of secreted proteins in a large patient cohort">
<meta property="og:description" content="Clinical relavance of secreted proteins in a large patient cohort.
">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SecAct</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Spatial transcriptomcis (ST) data</h6>
    <a class="dropdown-item" href="../articles/stPattern.html">Signaling patterns and velocities for multi-cellular ST data</a>
    <a class="dropdown-item" href="../articles/stCCC.html">Intercellular communication for single-cell resolution ST data</a>
    <h6 class="dropdown-header" data-toc-skip>Single-cell RNA sequencing data</h6>
    <a class="dropdown-item" href="../articles/scCCC.html">Cell-cell communication mediated by secreted proteins</a>
    <a class="dropdown-item" href="../articles/scState.html">Secreted protein signaling activity for various cell states</a>
    <h6 class="dropdown-header" data-toc-skip>Bulk RNA sequencing data</h6>
    <a class="dropdown-item" href="../articles/bulkDiff.html">Secreted protein activity difference between two phenotypes</a>
    <a class="dropdown-item" href="../articles/bulkCohort.html">Clinical relavance of secreted proteins in a large patient cohort</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/data2intelligence/SecAct/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Clinical relavance of secreted proteins in a large patient cohort</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/data2intelligence/SecAct/blob/HEAD/vignettes/bulkCohort.Rmd" class="external-link"><code>vignettes/bulkCohort.Rmd</code></a></small>
      <div class="d-none name"><code>bulkCohort.Rmd</code></div>
    </div>

    
    
<p>This tutorial demonstrates how to infer the signaling activity of
secreted proteins from a large cohort, and then calculate their activity
relevance with clinical data. The input expression values, from either
RNA sequencing or MicroArray, should be transformed by log2(x+1). Of
note, x could be FPKM, RPKM, or TPM for RNA-Seq data.</p>
<p>For this demonstration, we employ an immunotherapy
<a href="https://github.com/ParkerICI/prince-trial-data" target="_blank" class="external-link">cohort</a>
of pancreatic cancer patients. Using the SecAct framework, we will
identify secreted proteins associated with immunotherapy efficacy.</p>
<div class="section level2">
<h2 id="prepare-expression-data">Prepare expression data<a class="anchor" aria-label="anchor" href="#prepare-expression-data"></a>
</h2>
<p>This section shows how to read an expression matrix to R
environment.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/data2intelligence/SecAct" class="external-link">SecAct</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># prepare expression matrix</span></span>
<span><span class="va">dataPath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"SecAct"</span><span class="op">)</span>, <span class="st">"extdata/"</span><span class="op">)</span></span>
<span><span class="va">expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dataPath</span>,<span class="st">"Pancreatic_Nivolumab_Padron2022.logTPM.gz"</span><span class="op">)</span>, check.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="infer-secreted-protein-activity">Infer secreted protein activity<a class="anchor" aria-label="anchor" href="#infer-secreted-protein-activity"></a>
</h2>
<p>We use <code>SecAct.activity.inference</code> to infer the activity
of &gt;1000 secreted proteins across patient samples.</p>
<p>If your data have control samples (e.g., normal patients), you can
assign them to <code>inputProfile_control</code>. <code>SecAct</code>
would normalize each gene’s expression values with the control samples,
i.e., substracting the mean gene expression of control samples.</p>
<p>If not available, just set <code>inputProfile_control=NULL</code>
(default). <code>SecAct</code> would normalize the expression values of
each gene to zero mean across all input profiles. In other words, the
mean gene expression of all input profiles are regarded as control.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># infer activity; ~10 mins</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SecAct.activity.inference.html">SecAct.activity.inference</a></span><span class="op">(</span>inputProfile<span class="op">=</span><span class="va">expr</span>, inputProfile_control<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># res$zscore stores activity</span></span>
<span><span class="va">act</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">zscore</span></span>
<span></span>
<span><span class="co"># show activity</span></span>
<span><span class="va">act</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span></span>
<span><span class="co">##                  2          3           6</span></span>
<span><span class="co">## A1BG    -27.104244 -24.572367 -13.2677965</span></span>
<span><span class="co">## A2M       6.680395   6.097205  14.8814865</span></span>
<span><span class="co">## A2ML1     4.369943  -2.016608   5.2873155</span></span>
<span><span class="co">## AADACL2  -1.193300  -2.046660  -0.1879382</span></span>
<span><span class="co">## ABHD15  -14.441596 -12.734870  -8.2587602</span></span>
<span><span class="co">## ABI3BP   20.304553  14.203647  25.4980024</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="calculate-clinical-relevance">Calculate clinical relevance<a class="anchor" aria-label="anchor" href="#calculate-clinical-relevance"></a>
</h2>
<p>We next calculate the risk score z for each secreted protein by
linking their activity levels with clinical data. For the clinical file,
please make sure the name of the 1st and 2nd columns are “Time” and
“Event”, respectively. <code>SecAct.coxph</code> will conduct Cox
proportional hazard (PH) regression, with gene risk scores represented
as z-scores (Coef / StdErr) from the two-sided Wald test. For each
endpoint analysis, clinical covariates (e.g., age, gender, stage) were
considered in regression if available.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read clinical infomation</span></span>
<span><span class="va">clinical</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dataPath</span>,<span class="st">"Pancreatic_Nivolumab_Padron2022.OS_Nivo+Sotiga+Chemo"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##    Time Event Age Gender ECOG</span></span>
<span><span class="co">## 8   834     0  55      1    0</span></span>
<span><span class="co">## 13   28     1  60      0    1</span></span>
<span><span class="co">## 15  273     1  69      0    0</span></span>
<span><span class="co">## 22  159     1  69      0    1</span></span>
<span><span class="co">## 24  569     0  62      0    0</span></span>
<span><span class="co">## 25  264     1  66      1    1</span></span>
<span></span>
<span><span class="co"># compute risk score</span></span>
<span><span class="va">riskMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SecAct.coxph.html">SecAct.coxph</a></span><span class="op">(</span><span class="va">act</span>, <span class="va">clinical</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">riskMat</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##         risk score z     p value</span></span>
<span><span class="co">## A1BG       -2.187803 0.028683943</span></span>
<span><span class="co">## A2M        -3.372297 0.000745439</span></span>
<span><span class="co">## A2ML1       1.668025 0.095310824</span></span>
<span><span class="co">## AADACL2    -1.960305 0.049960191</span></span>
<span><span class="co">## ABHD15     -2.511890 0.012008638</span></span>
<span><span class="co">## ABI3BP     -1.679394 0.093075338</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualize-risk-score">Visualize risk score<a class="anchor" aria-label="anchor" href="#visualize-risk-score"></a>
</h2>
<p>User can visualize any secreted proteins of interest using
<code>SecAct.lollipop.plot</code>. Here, we select the top high and low
risk secreted proteins (SPs), which are associated with unfavorable and
favorable clinical outcomes, respectively.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select the top secreted proteins</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span><span class="va">high.risk.SPs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">riskMat</span><span class="op">[</span>,<span class="st">"risk score z"</span><span class="op">]</span>,decreasing<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n</span><span class="op">]</span></span>
<span><span class="va">low.risk.SPs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">riskMat</span><span class="op">[</span>,<span class="st">"risk score z"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n</span><span class="op">]</span></span>
<span></span>
<span><span class="va">SPs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">high.risk.SPs</span>, <span class="va">low.risk.SPs</span><span class="op">)</span></span>
<span><span class="co"># user can assign any secreted proteins to SPs</span></span>
<span></span>
<span><span class="co"># generate a vector</span></span>
<span><span class="va">fg.vec</span> <span class="op">&lt;-</span> <span class="va">riskMat</span><span class="op">[</span><span class="va">SPs</span>,<span class="st">"risk score z"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># visualize in lollipop plot</span></span>
<span><span class="fu"><a href="../reference/SecAct.lollipop.plot.html">SecAct.lollipop.plot</a></span><span class="op">(</span><span class="va">fg.vec</span>, title<span class="op">=</span><span class="st">"Risk score z"</span><span class="op">)</span></span></code></pre></div>
<p><img src="img/bulk_RiskScore.png" width="50%" style="border:none"></p>
<p>A positive value means the secreted protein is associated with worse
survival while a negative value means the secreted protein is associated
with better survival.</p>
</div>
<div class="section level2">
<h2 id="draw-survival-plot">Draw survival plot<a class="anchor" aria-label="anchor" href="#draw-survival-plot"></a>
</h2>
<p>User can select a secreted protein of interest to draw its survival
curve using <code>SecAct.survival.plot</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Take secreted protein WNT7B as an example</span></span>
<span><span class="va">riskMat</span><span class="op">[</span><span class="st">"WNT7B"</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co">## risk score z      p value </span></span>
<span><span class="co">## 3.3252642698 0.0008833476 </span></span>
<span></span>
<span><span class="co"># visualize in survival plot</span></span>
<span><span class="fu"><a href="../reference/SecAct.survival.plot.html">SecAct.survival.plot</a></span><span class="op">(</span><span class="va">act</span>, <span class="va">clinical</span>, <span class="st">"WNT7B"</span>, x.title<span class="op">=</span><span class="st">"Overall (Days)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="img/bulk_Survival.png" width="50%" style="border:none"></p>
<p>The activity cutoff was selected by maximizing the difference between
high and low activity patient groups.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Beibei Ru, Peng Jiang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
