<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Secreted protein signaling activity change between two phenotypes â€¢ SecAct</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Secreted protein signaling activity change between two phenotypes">
<meta name="description" content="Secreted protein signaling activity change between two phenotypes.
">
<meta property="og:description" content="Secreted protein signaling activity change between two phenotypes.
">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SecAct</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fas fa-home fa-lg"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials">
<li><h6 class="dropdown-header" data-toc-skip>Spatial transcriptomcis (ST) data</h6></li>
    <li><a class="dropdown-item" href="../articles/stPattern.html">Signaling patterns and velocities for multi-cellular ST data</a></li>
    <li><a class="dropdown-item" href="../articles/stCCC.html">Intercellular communication for single-cell resolution ST data</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Single-cell RNA sequencing data</h6></li>
    <li><a class="dropdown-item" href="../articles/scCCC.html">Cell-cell communication mediated by secreted proteins</a></li>
    <li><a class="dropdown-item" href="../articles/scState.html">Secreted protein signaling activity for various cell states</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bulk RNA sequencing data</h6></li>
    <li><a class="dropdown-item" href="../articles/bulkChange.html">Secreted protein activity change between two phenotypes</a></li>
    <li><a class="dropdown-item" href="../articles/bulkCohort.html">Clinical relevance of secreted proteins in a large patient cohort</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/data2intelligence/SecAct/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Secreted protein signaling activity change between two phenotypes</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/data2intelligence/SecAct/blob/HEAD/vignettes/bulkChange.Rmd" class="external-link"><code>vignettes/bulkChange.Rmd</code></a></small>
      <div class="d-none name"><code>bulkChange.Rmd</code></div>
    </div>

    
    
<p>This tutorial demonstrates how to infer the activity change of
secreted proteins between two phenotypes, such as treatment vs control,
tumor vs normal, or responder vs non-responder. The input expression
values, from either RNA sequencing or MicroArray, should be transformed
by log2(x+1). Of note, x could be FPKM, RPKM, or TPM for RNA-Seq
data.</p>
<p>For this demonstration, we employ a clinical
<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE100093" target="_blank" class="external-link">cohort</a>
of patients with discoid lupus erythematosus, who received a single dose
of anti-IFNG antibody AMG 811. Using the SecAct framework, we will
predict the activity change of secreted proteins between treatment and
control groups.</p>
<div class="section level2">
<h2 id="prepare-expression-data">Prepare expression data<a class="anchor" aria-label="anchor" href="#prepare-expression-data"></a>
</h2>
<p>This section shows how to read an expression matrix to the R
environment and separate samples into the treatment and control
groups.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/data2intelligence/SecAct" class="external-link">SecAct</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># prepare expression matrix</span></span>
<span><span class="va">dataPath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"SecAct"</span><span class="op">)</span>, <span class="st">"extdata/"</span><span class="op">)</span></span>
<span><span class="va">expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dataPath</span>,<span class="st">"GSE100093.IFNG.expr.gz"</span><span class="op">)</span>, sep<span class="op">=</span><span class="st">"\t"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># read meta infomation</span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dataPath</span>,<span class="st">"GSE100093.IFNG.meta"</span><span class="op">)</span>, sep<span class="op">=</span><span class="st">"\t"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset treatment and control groups</span></span>
<span><span class="va">expr.treatment</span> <span class="op">&lt;-</span> <span class="va">expr</span><span class="op">[</span>, <span class="va">meta</span><span class="op">[</span>,<span class="st">"Treatment"</span><span class="op">]</span><span class="op">==</span><span class="st">"Anti-IFNG"</span><span class="op">]</span></span>
<span><span class="va">expr.control</span> <span class="op">&lt;-</span> <span class="va">expr</span><span class="op">[</span>, <span class="va">meta</span><span class="op">[</span>,<span class="st">"Treatment"</span><span class="op">]</span><span class="op">==</span><span class="st">"Control"</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="infer-activity-change">Infer activity change<a class="anchor" aria-label="anchor" href="#infer-activity-change"></a>
</h2>
<p>We input <code>expr.treatment</code> and <code>expr.control</code>
into <code>SecAct.activity.inference</code> to infer the activity change
of secreted proteins between treatment and control groups.
<code>SecAct</code> calculates the OVERALL activity change between two
phenotypes with default setting
<code>is.singleSampleLevel = FALSE</code>. You can also obtain the
activity change for each SINGLE sample in <code>expr.treatment</code> by
setting <code>is.singleSampleLevel = TRUE</code>. Of note, if your
samples are paired, just change the default setting
<code>is.paired = FALSE</code> to <code>TRUE</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># infer activity change; ~2 mins</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SecAct.activity.inference.html">SecAct.activity.inference</a></span><span class="op">(</span></span>
<span>  inputProfile <span class="op">=</span> <span class="va">expr.treatment</span>, </span>
<span>  inputProfile_control <span class="op">=</span> <span class="va">expr.control</span>,</span>
<span>  is.singleSampleLevel <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  is.paired <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># res$zscore stores activity change</span></span>
<span><span class="va">act.change</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">zscore</span></span>
<span></span>
<span><span class="co"># show activity change</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">act.change</span><span class="op">)</span></span></code></pre></div>
<p>A positive value means the secreted protein has high activity in the
treatment group while a negative value means the secreted protein has
high activity in the control group.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show IFNG activity change</span></span>
<span><span class="va">act.change</span><span class="op">[</span><span class="st">"IFNG"</span>,<span class="st">"Change"</span><span class="op">]</span></span></code></pre></div>
<p>We can see that IFNG activity was reduced as expected, consistent
with the treatment group receiving anti-IFNG therapy.</p>
</div>
<div class="section level2">
<h2 id="visualize-activity-change">Visualize activity change<a class="anchor" aria-label="anchor" href="#visualize-activity-change"></a>
</h2>
<p>User can visualize any secreted proteins of interest using
<code>SecAct.bar.plot</code>. Here, we select the top up- and
down-regulated secreted proteins (SPs) between treatment and control as
examples.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select the top secreted proteins</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span><span class="va">up.SPs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">act.change</span><span class="op">[</span>,<span class="st">"Change"</span><span class="op">]</span>,decreasing<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n</span><span class="op">]</span></span>
<span><span class="va">down.SPs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">act.change</span><span class="op">[</span>,<span class="st">"Change"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n</span><span class="op">]</span></span>
<span></span>
<span><span class="va">SPs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">up.SPs</span>, <span class="va">down.SPs</span><span class="op">)</span></span>
<span><span class="co"># user can assign any secreted proteins to SPs</span></span>
<span></span>
<span><span class="co"># generate a vector</span></span>
<span><span class="va">fg.vec</span> <span class="op">&lt;-</span> <span class="va">act.change</span><span class="op">[</span><span class="va">SPs</span>,<span class="st">"Change"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># visualize in bar plot</span></span>
<span><span class="fu"><a href="../reference/SecAct.bar.plot.html">SecAct.bar.plot</a></span><span class="op">(</span><span class="va">fg.vec</span>, title<span class="op">=</span><span class="st">"Activity Change (treatment vs control)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="img/bulk_ActChange.png" width="50%" style="border:none"></p>
<p>A positive value means the secreted protein has high activity in the
treatment group while a negative value means the secreted protein has
high activity in the control group.</p>
</div>
<div class="section level2">
<h2 id="run-a-differential-profile">Run a differential profile<a class="anchor" aria-label="anchor" href="#run-a-differential-profile"></a>
</h2>
<p>If you already have a differential profile (or signature) instead of
the expression matrix from two phenotypes, you can directly run
<code>SecAct.activity.inference</code> with
<code>is.differential = TRUE</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># compare two phenotypes to generate a differential profile</span></span>
<span><span class="va">expr.diff</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">expr.treatment</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">expr.control</span><span class="op">)</span>, </span>
<span>  ncol <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    row_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">expr.treatment</span><span class="op">)</span>,</span>
<span>    col_names <span class="op">=</span> <span class="st">"Diff"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run a differential profile; ~2 mins</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SecAct.activity.inference.html">SecAct.activity.inference</a></span><span class="op">(</span></span>
<span>  inputProfile <span class="op">=</span> <span class="va">expr.diff</span>, </span>
<span>  is.differential <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># res$zscore stores activity change</span></span>
<span><span class="va">act.change</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">zscore</span></span>
<span></span>
<span><span class="co"># show activity change</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">act.change</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Beibei Ru.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
