<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Cell-cell communication for single-cell resolution spatial transcriptomics data.
">
<title>Cell-cell communication for single-cell resolution spatial transcriptomics data â€¢ SecAct</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Cell-cell communication for single-cell resolution spatial transcriptomics data">
<meta property="og:description" content="Cell-cell communication for single-cell resolution spatial transcriptomics data.
">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SecAct</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Spatial transcriptomcis (ST) data</h6>
    <a class="dropdown-item" href="../articles/stPattern.html">Signaling patterns and velocities for multi-cellular ST data</a>
    <a class="dropdown-item" href="../articles/stCCC.html">Intercellular communication for single-cell resolution ST data</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single-cell RNA sequencing data</h6>
    <a class="dropdown-item" href="../articles/scCCC.html">Cell-cell communication mediated by secreted proteins</a>
    <a class="dropdown-item" href="../articles/scState.html">Secreted protein signaling activity for various cell states</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Bulk RNA sequencing data</h6>
    <a class="dropdown-item" href="../articles/bulkChange.html">Secreted protein activity change between two phenotypes</a>
    <a class="dropdown-item" href="../articles/bulkCohort.html">Clinical relevance of secreted proteins in a large patient cohort</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/data2intelligence/SecAct/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Cell-cell communication for single-cell resolution spatial transcriptomics data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/data2intelligence/SecAct/blob/HEAD/vignettes/stCCC.Rmd" class="external-link"><code>vignettes/stCCC.Rmd</code></a></small>
      <div class="d-none name"><code>stCCC.Rmd</code></div>
    </div>

    
    
<p>This tutorial demonstrates how to infer cell-cell communication (CCC)
mediated by secreted proteins from single-cell resolution spatial
transcriptomics (ST) data. For this demonstration, we employ a liver
cancer ST sample from CosMx platform
[<a href="https://nanostring.com/products/cosmx-spatial-molecular-imager/ffpe-dataset/human-liver-rna-ffpe-dataset/" target="_blank" class="external-link">Link</a>].
Before running the tutorial, make sure that you have installed SecAct as
well as our previous R package <a href="https://github.com/data2intelligence/SpaCET/" target="_blank" class="external-link">SpaCET</a>. Here, SpaCET will be used to create a SpaCET
object to store the ST data.</p>
<div class="section level2">
<h2 id="read-st-data-to-a-spacet-object">Read ST data to a SpaCET object<a class="anchor" aria-label="anchor" href="#read-st-data-to-a-spacet-object"></a>
</h2>
<p>To load data into R, user can create a SpaCET object by preparing
three types of input data.</p>
<ol style="list-style-type: decimal">
<li>count data. The spatial transcriptomics count data must be in the
format of matrix with gene name (row) x cell ID (column).</li>
<li>spatial location information. The cell coordinates should be in the
format of matrix with cell ID (row) x coordinates (column). The 1st and
2nd columns represent X and Y coordinates, respectively.</li>
<li>meta data (optional). The meta data should be in the format of
matrix with cell ID (row) and meta annotation (column).</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/data2intelligence/SecAct" class="external-link">SecAct</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/data2intelligence/SpaCET" class="external-link">SpaCET</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># download data from https://hpc.nih.gov/~Jiang_Lab/SecAct_Package/LIHC_CosMx_data.rda</span></span>
<span><span class="co"># load CosMx data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"LIHC_CosMx_data.rda"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show count matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="co">## [1]   1000 460441</span></span>
<span></span>
<span><span class="co"># show coordinate matrix</span></span>
<span><span class="va">spotCoordinates</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co">##             x_slide_mm y_slide_mm</span></span>
<span><span class="co">## c_2_100_733    6.70816    9.03340</span></span>
<span><span class="co">## c_2_101_240    7.50024    8.86156</span></span>
<span><span class="co">## c_2_101_339    7.60560    8.73604</span></span>
<span><span class="co">## c_2_101_452    7.62984    9.17260</span></span>
<span><span class="co">## c_2_102_179    7.86824    8.99368</span></span>
<span></span>
<span><span class="co"># show meta information</span></span>
<span><span class="va">metaData</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co">##                    cellType         niche</span></span>
<span><span class="co">## c_2_100_733      Tumor_core tumor subtype</span></span>
<span><span class="co">## c_2_101_240 Tumor_interface     interface</span></span>
<span><span class="co">## c_2_101_339      Tumor_core         tumor</span></span>
<span><span class="co">## c_2_101_452      Tumor_core tumor subtype</span></span>
<span><span class="co">## c_2_102_179      Tumor_core         tumor</span></span>
<span></span>
<span></span>
<span><span class="co"># create a SpaCET object</span></span>
<span><span class="va">SpaCET_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpaCET/man/create.SpaCET.object.html" class="external-link">create.SpaCET.object</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="va">counts</span>,</span>
<span>  spotCoordinates <span class="op">=</span> <span class="va">spotCoordinates</span>,</span>
<span>  metaData <span class="op">=</span> <span class="va">metaData</span>,</span>
<span>  imagePath <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  platform <span class="op">=</span> <span class="st">"CosMx"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show this object.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">SpaCET_obj</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># filter out cells with less than 50 expressed genes</span></span>
<span><span class="va">SpaCET_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpaCET/man/SpaCET.quality.control.html" class="external-link">SpaCET.quality.control</a></span><span class="op">(</span><span class="va">SpaCET_obj</span>, min.genes<span class="op">=</span><span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## [1] "Removing cells with less than 50 expressed genes."</span></span>
<span><span class="co">## [1] "16926 cells are removed."</span></span>
<span><span class="co">## [1] "443515 cells are kept."</span></span>
<span></span>
<span><span class="co"># plot the QC metrics</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpaCET/man/SpaCET.visualize.spatialFeature.html" class="external-link">SpaCET.visualize.spatialFeature</a></span><span class="op">(</span></span>
<span>  <span class="va">SpaCET_obj</span>, </span>
<span>  spatialType <span class="op">=</span> <span class="st">"QualityControl"</span>, </span>
<span>  spatialFeatures<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"UMI"</span>,<span class="st">"Gene"</span><span class="op">)</span>,</span>
<span>  imageBg <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  pointSize <span class="op">=</span> <span class="fl">0.1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="img/HCC_CoxMx_QC.png" width="100%" style="border:none"></p>
<p>The <code>SpaCET_obj@metaData</code> contains the cell type
annotation from the original study.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show the spatial distribution of all cell types</span></span>
<span><span class="va">my_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">'B'</span><span class="op">=</span><span class="st">'#C88888'</span>,<span class="st">'Erythrocyte'</span><span class="op">=</span><span class="st">'#fe666d'</span>,<span class="st">'T.alpha.beta'</span><span class="op">=</span><span class="st">'#B95FBB'</span>,<span class="st">'T.gamma.delta'</span><span class="op">=</span><span class="st">'#3288bd'</span>,</span>
<span>  <span class="st">'NK'</span><span class="op">=</span><span class="st">'#bb8761'</span>,<span class="st">'Hepatocyte'</span><span class="op">=</span><span class="st">'#63636d'</span>,<span class="st">'Cholangiocyte'</span><span class="op">=</span><span class="st">'#de77ae'</span>,<span class="st">'Endothelial'</span><span class="op">=</span><span class="st">'#D4D915'</span>,</span>
<span>  <span class="st">'Fibroblast'</span><span class="op">=</span><span class="st">'#66c2a5'</span>,<span class="st">'Macrophage'</span><span class="op">=</span><span class="st">'#ff9a36'</span>,<span class="st">'Tumor_core'</span><span class="op">=</span><span class="st">'#A4DFF2'</span>,<span class="st">'Tumor_boundary'</span><span class="op">=</span><span class="st">'blue'</span>,</span>
<span>  <span class="st">'Other'</span><span class="op">=</span><span class="st">'#cccccc'</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpaCET/man/SpaCET.visualize.spatialFeature.html" class="external-link">SpaCET.visualize.spatialFeature</a></span><span class="op">(</span></span>
<span>  <span class="va">SpaCET_obj</span>, </span>
<span>  spatialType <span class="op">=</span> <span class="st">"metaData"</span>, </span>
<span>  spatialFeatures<span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  colors <span class="op">=</span> <span class="va">my_cols</span>,</span>
<span>  pointSize <span class="op">=</span> <span class="fl">0.1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="img/HCC_CoxMx_cellType.png" width="55%" style="border:none"></p>
</div>
<div class="section level2">
<h2 id="infer-secreted-protein-activity">Infer secreted protein activity<a class="anchor" aria-label="anchor" href="#infer-secreted-protein-activity"></a>
</h2>
<p>After loading ST data, user can run
<code>SecAct.activity.inference.ST</code> to infer the signaling
activities of secreted proteins for each spot. The output are stored in
<code>SpaCET_obj @results $SecAct_output $SecretedProteinActivity</code>,
which includes four items, (1) beta: regression coefficients; (2) se:
standard error; (3) zscore: beta/se; (4): pvalue: two-sided test p value
of z score from permutation test.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># infer activity; 60 minutes</span></span>
<span><span class="va">SpaCET_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SecAct.activity.inference.ST.html">SecAct.activity.inference.ST</a></span><span class="op">(</span></span>
<span>  inputProfile <span class="op">=</span> <span class="va">SpaCET_obj</span>,</span>
<span>  scale.factor <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  sigFilter<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># The running times mentioned here and onwards were obtained</span></span>
<span><span class="co"># from our HPC with 20 cores and 200 GB of memory.</span></span>
<span></span>
<span><span class="co"># show activity</span></span>
<span><span class="va">SpaCET_obj</span> <span class="op">@</span><span class="va">results</span> <span class="op">$</span><span class="va">SecAct_output</span> <span class="op">$</span><span class="va">SecretedProteinActivity</span> <span class="op">$</span><span class="va">zscore</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="infer-cell-cell-communication">Infer cell-cell communication<a class="anchor" aria-label="anchor" href="#infer-cell-cell-communication"></a>
</h2>
<p>After calculating the secreted protein activity, SecAct could further
estimate the consensus pattern from these inferred signaling activities
across the whole tissue slide. This module contains two steps.</p>
<p>First, <code>SecAct</code> filters 1248 secreted proteins to identify
the significant secreted proteins mediating intercellular communication
in this slide. To achieve this, <code>SecAct</code> will calculate the
Spearman correlation of secreted protein s signaling activity and its
neighbor cellsâ€™ sum expression. The p values were adjusted by the
Benjamini-Hochberg (BH) method as false discovery rate (FDR). The
cutoffs are r &gt; 0.05 and FDR &lt; 0.01.</p>
<p>Second, We designed a computational strategy to statistically test
whether a type of secreted protein s mediates the intercellular
communication from cell type 1 to cell type 2. Cell-type annotations
were retrieved from the original study. First, we identified all
neighboring pairs of cells, where one cell was from cell type 1 and the
other from cell type 2, based on spatial proximity (distance &lt; 20
Î¼m). Among these neighboring pairs, we identified communicating pairs of
cells as those where cell type 1 expressed RNA for s (count &gt; 0), and
cell type 2 exhibited signaling activity for s (activity score &gt; 0).
We then calculated a communication score for cell type 1 â†’ s â†’ cell type
2, defined as the ratio of communicating pairs to total neighboring
pairs. The p-value is Probability (random ratio &gt;= real ratio),
computed through 1000 randomizations adjusted by the Benjaminiâ€“Hochberg
procedure. In each randomization, we randomly permuted the cell IDs
within each cell type. A cell type 1 â†’ s â†’ cell type 2 communication is
considered significant if ratio &gt; 0.2 and FDR &lt; 0.01.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># infer CCC; 60 minutes</span></span>
<span><span class="va">SpaCET_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SecAct.CCC.scST.html">SecAct.CCC.scST</a></span><span class="op">(</span></span>
<span>    <span class="va">SpaCET_obj</span>,</span>
<span>    cellType_meta <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>    scale.factor <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>    radius <span class="op">=</span> <span class="fl">0.02</span>,</span>
<span>    ratio_cutoff <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>    padj_cutoff <span class="op">=</span> <span class="fl">0.01</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">SpaCET_obj</span> <span class="op">@</span><span class="va">results</span> <span class="op">$</span><span class="va">SecAct_output</span> <span class="op">$</span><span class="va">SecretedProteinCCC</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualize-cell-cell-communication">Visualize cell-cell communication<a class="anchor" aria-label="anchor" href="#visualize-cell-cell-communication"></a>
</h2>
<p>We provide two types of visualization plot for cell-cell
communication, i.e., heatmap and circle plot. The number in the heatmap
represents the count of secreted proteins from senders to receivers.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/SecAct.CCC.heatmap.html">SecAct.CCC.heatmap</a></span><span class="op">(</span><span class="va">SpaCET_obj</span>, row.sorted<span class="op">=</span><span class="cn">TRUE</span>, column.sorted<span class="op">=</span><span class="cn">TRUE</span>, colors_cellType<span class="op">=</span><span class="va">my_cols</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/SecAct.CCC.circle.html">SecAct.CCC.circle</a></span><span class="op">(</span><span class="va">SpaCET_obj</span>, colors_cellType<span class="op">=</span><span class="va">my_cols</span><span class="op">)</span></span></code></pre></div>
<p><img src="img/HCC_CoxMx_heatmap.png" width="50%" style="border:none"><img src="img/HCC_CoxMx_circlize.png" width="45%" style="border:none"></p>
<p>User can select the cell-cell communication of interest from
<code>SpaCET_obj @results $SecAct_output $SecretedProteinCCC</code> to
visualize them using <code>SecAct.CCC.dot</code>. You need to assign
values to <code>sender</code>, <code>secretedProtein</code>, and
<code>receiver</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellTypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Tumor_boundary"</span>,<span class="st">"Fibroblast"</span>,<span class="st">"Macrophage"</span>,<span class="st">"Endothelial"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">secretedProtein</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"BGN"</span>,<span class="st">"COL1A1"</span>,<span class="st">"COL1A2"</span>,<span class="st">"DCN"</span>,<span class="st">"IGFBP5"</span>,</span>
<span>  <span class="st">"LGALS1"</span>,<span class="st">"LGALS9"</span>,<span class="st">"LYZ"</span>,<span class="st">"LUM"</span>,<span class="st">"MGP"</span>,</span>
<span>  <span class="st">"SPP1"</span>,<span class="st">"THBS1"</span>,<span class="st">"THBS2"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/SecAct.CCC.dot.html">SecAct.CCC.dot</a></span><span class="op">(</span></span>
<span>  <span class="va">SpaCET_obj</span>, </span>
<span>  colors_cellType<span class="op">=</span><span class="va">my_cols</span>, </span>
<span>  sender<span class="op">=</span><span class="va">cellTypes</span>, </span>
<span>  secretedProtein<span class="op">=</span><span class="va">secretedProtein</span>, </span>
<span>  receiver<span class="op">=</span><span class="va">cellTypes</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="img/HCC_CoxMx_dot.png" width="60%" style="border:none"></p>
</div>
<div class="section level2">
<h2 id="visualize-signaling-velocity">Visualize signaling velocity<a class="anchor" aria-label="anchor" href="#visualize-signaling-velocity"></a>
</h2>
<p>User can select the cell-cell communication of interest from
<code>SpaCET_obj @results $SecAct_output $SecretedProteinCCC</code> to
visualize them using <code>SecAct.signaling.velocity.scST</code>. You
need to assign values to <code>sender</code>,
<code>secretedProtein</code>, and <code>receiver</code>.</p>
<pre><code><span><span class="fu"><a href="../reference/SecAct.signaling.velocity.scST.html">SecAct.signaling.velocity.scST</a></span><span class="op">(</span></span>
<span>  <span class="va">SpaCET_obj</span>, </span>
<span>  sender<span class="op">=</span><span class="st">"Fibroblast"</span>, </span>
<span>  secretedProtein<span class="op">=</span><span class="st">"THBS2"</span>, </span>
<span>  receiver<span class="op">=</span><span class="st">"Tumor_boundary"</span>, </span>
<span>  cellType_meta<span class="op">=</span><span class="st">"cellType"</span></span>
<span><span class="op">)</span></span></code></pre>
<p><img src="img/HCC_CoxMx_velocity.png" width="60%" style="border:none"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Beibei Ru, Peng Jiang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
