<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Signaling patterns and velocities for multi-cellular spatial transcriptomics data.
">
<title>Signaling patterns and velocities for multi-cellular spatial transcriptomics data • SecAct</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Signaling patterns and velocities for multi-cellular spatial transcriptomics data">
<meta property="og:description" content="Signaling patterns and velocities for multi-cellular spatial transcriptomics data.
">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SecAct</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Spatial transcriptomcis (ST) data</h6>
    <a class="dropdown-item" href="../articles/stPattern.html">Signaling patterns and velocities for multi-cellular ST data</a>
    <a class="dropdown-item" href="../articles/stCCC.html">Intercellular communication for single-cell resolution ST data</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single-cell RNA sequencing data</h6>
    <a class="dropdown-item" href="../articles/scCCC.html">Cell-cell communication mediated by secreted proteins</a>
    <a class="dropdown-item" href="../articles/scState.html">Secreted protein signaling activity for various cell states</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Bulk RNA sequencing data</h6>
    <a class="dropdown-item" href="../articles/bulkChange.html">Secreted protein activity change between two phenotypes</a>
    <a class="dropdown-item" href="../articles/bulkCohort.html">Clinical relevance of secreted proteins in a large patient cohort</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/data2intelligence/SecAct/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Signaling patterns and velocities for multi-cellular spatial transcriptomics data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/data2intelligence/SecAct/blob/HEAD/vignettes/stPattern.Rmd" class="external-link"><code>vignettes/stPattern.Rmd</code></a></small>
      <div class="d-none name"><code>stPattern.Rmd</code></div>
    </div>

    
    
<p>This tutorial demonstrates how to infer secreted protein activities
of each spot for a liver cancer spatial transcriptomics (ST) sample from
Visium platform (<a href="https://www.thno.org/v12p4163.htm" target="_blank" class="external-link">Paper</a>). Each spot is 55 µm in diameter covering <a href="https://kb.10xgenomics.com/hc/en-us/articles/360035487952-How-many-cells-are-captured-in-a-single-spot" target="_blank" class="external-link">1-10 cells</a>. Further, SecAct provides additional
modules to analyze the signaling patterns and velocities across the
whole slide. Before running the tutorial, make sure that you have
installed SecAct as well as our previous R package <a href="https://github.com/data2intelligence/SpaCET/" target="_blank" class="external-link">SpaCET</a>. Here, SpaCET will be used to create a SpaCET
object to store the ST data.</p>
<div class="section level2">
<h2 id="read-st-data-to-a-spacet-object">Read ST data to a SpaCET object<a class="anchor" aria-label="anchor" href="#read-st-data-to-a-spacet-object"></a>
</h2>
<p>To load data into R, user can create a SpaCET object by using
<code>create.SpaCET.object.10X</code>. Please make sure that
<code>visiumPath</code> points to the standard output folders of 10x
Space Ranger. If the ST data is not from 10x/Visium, you can use
<code>create.SpaCET.object</code> instead
[<a href="https://data2intelligence.github.io/SpaCET/articles/oldST_PDAC.html" target="_blank" class="external-link">details</a>].</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/data2intelligence/SecAct" class="external-link">SecAct</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/data2intelligence/SpaCET" class="external-link">SpaCET</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set the path to the data folder</span></span>
<span><span class="va">dataPath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"SecAct"</span><span class="op">)</span>, <span class="st">"extdata/"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load ST data to create an SpaCET object</span></span>
<span><span class="va">visiumPath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dataPath</span>,<span class="st">"Visium_HCC/"</span><span class="op">)</span></span>
<span><span class="va">SpaCET_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpaCET/man/create.SpaCET.object.10X.html" class="external-link">create.SpaCET.object.10X</a></span><span class="op">(</span>visiumPath <span class="op">=</span> <span class="va">visiumPath</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># filter out spots with less than 1000 expressed genes</span></span>
<span><span class="va">SpaCET_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpaCET/man/SpaCET.quality.control.html" class="external-link">SpaCET.quality.control</a></span><span class="op">(</span><span class="va">SpaCET_obj</span>, min.genes <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot the QC metrics</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpaCET/man/SpaCET.visualize.spatialFeature.html" class="external-link">SpaCET.visualize.spatialFeature</a></span><span class="op">(</span></span>
<span>  <span class="va">SpaCET_obj</span>, </span>
<span>  spatialType <span class="op">=</span> <span class="st">"QualityControl"</span>, </span>
<span>  spatialFeatures <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"UMI"</span>,<span class="st">"Gene"</span><span class="op">)</span>,</span>
<span>  imageBg <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="img/HCC_visium_QC.png" width="100%" style="border:none"></p>
</div>
<div class="section level2">
<h2 id="infer-secreted-protein-activity">Infer secreted protein activity<a class="anchor" aria-label="anchor" href="#infer-secreted-protein-activity"></a>
</h2>
<p>After loading ST data, user can run
<code>SecAct.activity.inference.ST</code> to infer the activities of
1248 secreted proteins for each spot. The output are stored in
<code>SpaCET_obj @results $SecAct_output $SecretedProteinActivity</code>,
which includes four items, (1) beta: regression coefficients; (2) se:
standard error; (3) zscore: beta/se; (4): pvalue: two-sided test p value
of z score from permutation test.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># infer activity; 25 minutes</span></span>
<span><span class="va">SpaCET_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SecAct.activity.inference.ST.html">SecAct.activity.inference.ST</a></span><span class="op">(</span></span>
<span>  inputProfile <span class="op">=</span> <span class="va">SpaCET_obj</span>,</span>
<span>  scale.factor <span class="op">=</span> <span class="fl">1e+05</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># The running times mentioned here and onwards were obtained</span></span>
<span><span class="co"># from our HPC with 20 cores and 200 GB of memory.</span></span>
<span></span>
<span><span class="co"># show activity</span></span>
<span><span class="va">SpaCET_obj</span> <span class="op">@</span><span class="va">results</span> <span class="op">$</span><span class="va">SecAct_output</span> <span class="op">$</span><span class="va">SecretedProteinActivity</span> <span class="op">$</span><span class="va">zscore</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span></span>
<span><span class="co">##           50x102      3x43     14x94</span></span>
<span><span class="co">## A1BG    4.179847  2.193798 -1.082260</span></span>
<span><span class="co">## A2M     1.990219  7.513577  2.520423</span></span>
<span><span class="co">## A2ML1   0.838544  0.298118  3.453744</span></span>
<span><span class="co">## AADACL2 1.790729 -3.226144  2.709726</span></span>
<span><span class="co">## ABHD15  3.224051 -1.534522  2.073572</span></span>
<span><span class="co">## ABI3BP  1.588963  6.592146 13.925150</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="estimate-signaling-pattern">Estimate signaling pattern<a class="anchor" aria-label="anchor" href="#estimate-signaling-pattern"></a>
</h2>
<p>After calculating the secreted protein activity, SecAct could further
estimate the consensus pattern from these inferred signaling activities
across the whole tissue slide. This module contains two steps.</p>
<p>First, <code>SecAct</code> filters 1248 secreted proteins to identify
the significant secreted proteins mediating intercellular communication
in this slide. To achieve this, <code>SecAct</code> will calculate the
Spearman correlation of spots’ signaling activity and spots’ neighbors’
RNA expression. The p values were adjusted by the Benjamini-Hochberg
(BH) method as false discovery rate (FDR). The cutoffs are r &gt; 0.05
and FDR &lt; 0.01.</p>
<p>Second, <code>SecAct</code> employs Non-negative Matrix Factorization
(<a href="https://cran.r-project.org/web/packages/NMF/vignettes/NMF-vignette.pdf" target="_blank" class="external-link">NMF</a>) to estimate the consensus signaling patterns. A
critical parameter in NMF is the factorization rank <code>k</code>. User
can assign a number vector to <code>k</code>, e.g., <code>k=2:5</code>.
Then, <code>SecAct.signaling.pattern</code> would find the optimal
number of factors determined as the point preceding the largest decrease
in the silhouette value. This will take a while. Based on our
pre-calculation, <code>k=3</code> is the optimal number of factors. To
save time, we directly run against <code>k=3</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># estimate signaling pattern; 3 minutes</span></span>
<span><span class="va">SpaCET_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SecAct.signaling.pattern.html">SecAct.signaling.pattern</a></span><span class="op">(</span><span class="va">SpaCET_obj</span>, k<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># plot signaling pattern</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpaCET/man/SpaCET.visualize.spatialFeature.html" class="external-link">SpaCET.visualize.spatialFeature</a></span><span class="op">(</span></span>
<span>    <span class="va">SpaCET_obj</span>, </span>
<span>    spatialType <span class="op">=</span> <span class="st">"SignalingPattern"</span>, </span>
<span>    spatialFeatures <span class="op">=</span> <span class="st">"All"</span>, </span>
<span>    imageBg <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    legend.position <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    colors<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#03c383"</span>,<span class="st">"#fbbf45"</span>,<span class="st">"#ef6a32"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="img/HCC_visium_pattern.png" width="100%" style="border:none"></p>
<p>Further, <code>SecAct</code> can identify secreted proteins
associated with each signaling pattern according to the matrix W from
NMF results. For one secreted protein (represented by a row in W), the
signaling pattern with a value at least twice as large as any other
pattern, is designated as the dominant pattern for that protein.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># identify secreted proteins dominated by pattern 3</span></span>
<span><span class="va">pattern.gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SecAct.signaling.pattern.gene.html">SecAct.signaling.pattern.gene</a></span><span class="op">(</span><span class="va">SpaCET_obj</span>, n<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show these genes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pattern.gene</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="calculate-signaling-velocity">Calculate signaling velocity<a class="anchor" aria-label="anchor" href="#calculate-signaling-velocity"></a>
</h2>
<p>Several secreted proteins with pattern 3 are related to
epithelial-mesenchymal transition process, such as COL1A1, TGFB1, and
SPARC. By integrating secreted protein-coding gene expression and
signaling activity, SecAct can also infer signaling velocity at each
spatial spot, indicating the direction and strength of secreted
signaling. Let’s take SPARC as an example.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show SPARC signaling velocity </span></span>
<span><span class="fu"><a href="../reference/SecAct.signaling.velocity.spotST.html">SecAct.signaling.velocity.spotST</a></span><span class="op">(</span><span class="va">SpaCET_obj</span>, gene <span class="op">=</span> <span class="st">"SPARC"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="img/HCC_visium_SPARC.png" width="100%" style="border:none"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show animated SPARC signaling velocity</span></span>
<span><span class="fu"><a href="../reference/SecAct.signaling.velocity.spotST.html">SecAct.signaling.velocity.spotST</a></span><span class="op">(</span><span class="va">SpaCET_obj</span>, gene <span class="op">=</span> <span class="st">"SPARC"</span>, animated<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> </span></code></pre></div>
<p><img src="img/HCC_visium_SPARC.gif" width="100%" style="border:none"></p>
</div>
<div class="section level2">
<h2 id="deconvolve-st-data">Deconvolve ST data<a class="anchor" aria-label="anchor" href="#deconvolve-st-data"></a>
</h2>
<p>For the current ST data, user also can run our previous R package
<code>SpaCET</code> to estimate the cell lineages for each spot. Click
<a href="https://data2intelligence.github.io/SpaCET/articles/visium_BC.html" target="_blank" class="external-link">here</a> for more details. Based on the deconvolution
results, we can see the interface region consists of fibroblasts,
macrophages, and endothelial cells.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># deconvolve ST data</span></span>
<span><span class="va">SpaCET_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpaCET/man/SpaCET.deconvolution.html" class="external-link">SpaCET.deconvolution</a></span><span class="op">(</span></span>
<span>  <span class="va">SpaCET_obj</span>, </span>
<span>  cancerType <span class="op">=</span> <span class="st">"LIHC"</span>, </span>
<span>  coreNo <span class="op">=</span> <span class="fl">8</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show the spatial distribution of all cell types.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpaCET/man/SpaCET.visualize.spatialFeature.html" class="external-link">SpaCET.visualize.spatialFeature</a></span><span class="op">(</span></span>
<span>  <span class="va">SpaCET_obj</span>, </span>
<span>  spatialType <span class="op">=</span> <span class="st">"CellFraction"</span>, </span>
<span>  spatialFeatures <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"Malignant"</span>,<span class="st">"CAF"</span>,<span class="st">"Endothelial"</span>,<span class="st">"Macrophage"</span>,</span>
<span>    <span class="st">"Hepatocyte"</span>,<span class="st">"B cell"</span>,<span class="st">"T CD4"</span>,<span class="st">"T CD8"</span><span class="op">)</span>, </span>
<span>  sameScaleForFraction <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  pointSize <span class="op">=</span> <span class="fl">0.1</span>, </span>
<span>  nrow <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="img/HCC_visium_deconvolve.png" width="100%" style="border:none"></p>
</div>
<div class="section level2">
<h2 id="calculate-hallmark-score">Calculate hallmark score<a class="anchor" aria-label="anchor" href="#calculate-hallmark-score"></a>
</h2>
<p>User also can run <code>SpaCET.GeneSetScore</code> to estimate the
hallmark scores for each spot. Click <a href="https://data2intelligence.github.io/SpaCET/articles/GeneSetScore.html" target="_blank" class="external-link">here</a> for more details. You can see that Pattern 3 is
correlated with epithelial-mesenchymal transition (EMT).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run gene set calculation</span></span>
<span><span class="va">SpaCET_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpaCET/man/SpaCET.GeneSetScore.html" class="external-link">SpaCET.GeneSetScore</a></span><span class="op">(</span><span class="va">SpaCET_obj</span>, GeneSets<span class="op">=</span><span class="st">"Hallmark"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># visualize EMT</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpaCET/man/SpaCET.visualize.spatialFeature.html" class="external-link">SpaCET.visualize.spatialFeature</a></span><span class="op">(</span></span>
<span>  <span class="va">SpaCET_obj</span>, </span>
<span>  spatialType <span class="op">=</span> <span class="st">"GeneSetScore"</span>, </span>
<span>  spatialFeatures <span class="op">=</span> <span class="st">"HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION"</span>,</span>
<span>  legend.position <span class="op">=</span> <span class="st">"right"</span>,</span>
<span>  imageBg<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>  pointSize<span class="op">=</span><span class="fl">1.2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="img/HCC_visium_EMT.png" width="100%" style="border:none"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Beibei Ru, Peng Jiang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
